### DISCUSSION TITLE
Is it possible to use inline raytracing?

### DISCUSSION THREAD

**Discussion by ohmaya** [o3de/o3de#14307] [Help Wanted]
Such as, using RT SRGs in non-RT .azsl files.

At least, RHI reports D3D12_RAYTRACING_TIER_1_1 for capable driver in:
`m_features.m_rayTracing = options5.RaytracingTier != D3D12_RAYTRACING_TIER_NOT_SUPPORTED;`

**Comment by ohmaya** [comment]
An error occurs as soon as I add the following to a graphics .azsl file:

```
#include <Atom/Features/SrgSemantics.azsli>

ShaderResourceGroup RayTracingGlobalSrg : SRG_RayTracingGlobal
{
    RaytracingAccelerationStructure m_scene;
}
```

**Can't use SRG_RayTracingGlobal.**

**https://docs.unity3d.com/2023.1/Documentation/ScriptReference/SystemInfo-supportsInlineRayTracing.html**

**Reply by siliconvoodoo** [reply]
That's maybe because the SRG at that slot already exists? What is the message?
Anyways to use RT the C++ side needs to kick the shader in a different environment than what it would be parameterized for the graphics purpose of the shader you are editing.

Notably those changes are:
- need a compute pipeline (not graphics)
- need library shader model profile shaders (lib_6x)
- need a TLAS and BLAS(es) to hit on anything
- need a global root signature
- need hitgroups local root signatures
- need a shader binding table (SBT) on the PSO that configures all that
- need a Dispatch kick (not a DrawCall)

Maybe check where and how o3de-atom-sampleviewer\Shaders\RHI\RayTracingClosestHitSolid.azsl is used?

**Reply by ohmaya** [reply]
@siliconvoodoo Could you point me to the AZSLc code that handles inline ray tracing? Thanks.

**Reply by siliconvoodoo** [reply]
There is no code that handles ray tracing in AZSLc. It takes the token of the input program as is, and re-emit then into an HLSL output.
AZSLc doesn't need to "understand" everything. Only a few structures, notably everything related to input.
So it registers symbols and types, and reflects all inputs into json by using --srg or --option command lines.
The raytracing (or compute or anything) is transparent to AZSLc, the code is just passed down.
Then DXC does the second step to DXIL or SPIRV.

**Reply by ohmaya** [reply]
@siliconvoodoo Your long reply describing the technial details disappears : )
